File: src/main/java/com/hazza/LeetCodeByJava/SegmentTwo/No99RecoverBinarySearchTree.java
package com.hazza.LeetCodeByJava.SegmentTwo;

import com.hazza.LeetCodeByJava.datastructure.TreeNode;

import java.util.Stack;

/**
 * Created with IntelliJ IDEA.
 * Description: https://leetcode.com/problems/recover-binary-search-tree/description/
 * Author: HazzaCheng
 * Contact: hazzacheng@gmail.com
 * Date: 2017-10-19
 * Time: 6:53 PM
 */
public class No99RecoverBinarySearchTree {
    private TreeNode firstNode = null, secondNode = null, prevNode = new TreeNode(Integer.MIN_VALUE);

    public void recoverTree(TreeNode root) {
        if (root == null) return;
        traverse(root);
        int temp = firstNode.val;
        firstNode.val = secondNode.val;
        secondNode.val = temp;
    }

    private void traverse(TreeNode root) {
        if (root == null) return;
        traverse(root.left);
        if (firstNode == null && prevNode.val >= root.val) firstNode = prevNode;
        if (firstNode != null && prevNode.val >= root.val) secondNode = root;
        prevNode = root;
        traverse(root.right);
    }
}

File: src/main/python/leetcode_by_python/tree/No99_Recover_Binary_Search_Tree.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Created by HazzaCheng on 2019-09-05
# https://leetcode.com/problems/recover-binary-search-tree/
import sys

from src.main.python.leetcode_by_python.data_structure.TreeNode import TreeNode


class Solution:
    # solution 1
    def __init__(self):
        self.firstNode, self.secondNode, self.prevNode = None, None, TreeNode(-sys.maxsize - 1)

    def recoverTree1(self, root: TreeNode) -> None:
        if not root:
            return
        self.firstNode, self.secondNode, self.prevNode = None, None, TreeNode(-sys.maxsize - 1)
        self.dfs(root)
        self.firstNode.val, self.secondNode.val = self.secondNode.val, self.firstNode.val

    def dfs(self, root):
        if not root:
            return
        self.dfs(root.left)

        if not self.firstNode and self.prevNode.val >= root.val:
            self.firstNode = self.prevNode
        if self.firstNode and self.prevNode.val >= root.val:
            self.secondNode = root
        self.prevNode = root

        self.dfs(root.right)

    # solution 2
    def recoverTree2(self, root: TreeNode) -> None:
        firstNode, secondNode, prevNode, tempNode = None, None, None, None

        while root:
            if root.left:
                tempNode = root.left
                while tempNode.right and tempNode.right != root:
                    tempNode = tempNode.right
                if tempNode.right:
                    if prevNode and prevNode.val >= root.val:
                        if not firstNode:
                            firstNode, secondNode = prevNode, root
                        else:
                            secondNode = root
                    prevNode = root
                    tempNode.right = None
                    root = root.right
                else:
                    tempNode.right = root
                    root = root.left
            else:
                if prevNode and prevNode.val >= root.val:
                    if not firstNode:
                        firstNode, secondNode = prevNode, root
                    else:
                        secondNode = root
                prevNode = root
                root = root.right

        if firstNode and secondNode:
            firstNode.val, secondNode.val = secondNode.val, firstNode.val
